services/QuillLMS/engines/evidence/lib/tasks/plagiarized_responses.rake

```diff
@@ -28,4 +28,89 @@ namespace :plagiarized_responses do
       end
     end
   end
+
+  desc "Reads a CSV of historical responses and applies theoretical plagiarism algorithms to them, then outputs the responses and those plagiarism results into a new CSV for comparison"
+
+  # Intended to operate on data generated via the following query:
+  # SELECT comprehension_prompts.text, entry, feedback_text, feedback_type, comprehension_plagiarism_texts.text AS plagiarism_text
+  # FROM feedback_histories
+  # JOIN comprehension_prompts
+  #     ON feedback_histories.prompt_id = comprehension_prompts.id
+  # JOIN comprehension_prompts_rules
+  #     ON feedback_histories.prompt_id = comprehension_prompts_rules.prompt_id
+  # JOIN comprehension_plagiarism_texts
+  #     ON comprehension_prompts_rules.rule_id = comprehension_plagiarism_texts.rule_id
+
+  task :test, [:input_csv_path] => :environment do |t, args|
+
+    # Levenshtein implementation lifted from https://stackoverflow.com/questions/16323571/measure-the-distance-between-two-strings-with-ruby
+    def levenshtein_distance(src, target, max = nil)
+      m = src.length
+      n = target.length
+      return m if n == 0
+      return n if m == 0
+      d = Array.new(m+1) {Array.new(n+1)}
+    
+      (0..m).each {|i| d[i][0] = i}
+      (0..n).each {|j| d[0][j] = j}
+      (1..n).each do |j|
+        (1..m).each do |i|
+          d[i][j] = if src[i-1] == target[j-1]  # adjust index into string
+                      d[i-1][j-1]       # no operation required
+                    else
+                      [
+                        d[i-1][j]+1,    # deletion
+                        d[i][j-1]+1,    # insertion
+                        d[i-1][j-1]+1  # substitution
+                      ].min
+                    end
+        end
+        return d[[j,m].min][j] if max && d[[j,m].min][j] >= max
+      end
+      d[m][n]
+    end
+
+    # Derived from lib/evidence/evidence/plagiarism_check.rb 'match_entry_on_passage'
+    def minimum_edit_distance_per_slice(entry, plagiarism_text)
+      match_minimum = 10
+
+      entry_arr = entry.gsub(/[[:punct:]]/, '').downcase.split
+      plagiarism_arr = plagiarism_text.gsub(/[[:punct:]]/, '').downcase.split
+
+      slices = entry_arr.each_cons(match_minimum).map { |slice| slice.join(' ') }
+      plagiarism_slices = plagiarism_arr.each_cons(match_minimum).map { |slice| slice.join(' ') }
+      min_distance = nil
+      slices.each do |slice|
+        plagiarism_slices.each do |pslice|
+          min_distance = levenshtein_distance(slice, pslice, 21)
+          return min_distance if min_distance <= 5
```

Comment:
integers should be created as variables so that they are not 'magic'

Reply:
Good call out.